src/debugging/DockerDebugConfigurationProvider.ts

```diff
@@ -111,27 +113,60 @@ export class DockerDebugConfigurationProvider implements DebugConfigurationProvi
     private async registerRemoveContainerAfterDebugging(resolvedConfiguration: ResolvedDebugConfiguration): Promise<void> {
         if (resolvedConfiguration.dockerOptions
             && (resolvedConfiguration.dockerOptions.removeContainerAfterDebug === undefined || resolvedConfiguration.dockerOptions.removeContainerAfterDebug)
-            && resolvedConfiguration.dockerOptions.containerNameToKill) {
+            && resolvedConfiguration.dockerOptions.containerName) {
 
             // Since Python is a special case as we handle waiting for the debugger to be ready while resolving
             // the launch configuration, and since this method comes later then we shouldn't remove a container
             // that we just created.
             // TODO: this needs to be removed as soon as the Python extension adds a way to retry while connecting to a remote debugger.
             if (resolvedConfiguration.type !== 'python') {
                 try {
-                    await this.dockerClient.removeContainer(resolvedConfiguration.dockerOptions.containerNameToKill, { force: true });
+                    await this.dockerClient.removeContainer(resolvedConfiguration.dockerOptions.containerName, { force: true });
                 } catch { }
             }
 
             // Now register the container for removal after the debug session ends
             const disposable = debug.onDidTerminateDebugSession(async session => {
                 const sessionConfiguration = <ResolvedDebugConfiguration>session.configuration;
 
-                if (sessionConfiguration
-                    && sessionConfiguration.dockerOptions
-                    && sessionConfiguration.dockerOptions.containerNameToKill === resolvedConfiguration.dockerOptions.containerNameToKill) {
+                if (sessionConfiguration?.dockerOptions?.containerName === resolvedConfiguration.dockerOptions.containerName) {
                     try {
-                        await this.dockerClient.removeContainer(resolvedConfiguration.dockerOptions.containerNameToKill, { force: true });
+                        await this.dockerClient.removeContainer(resolvedConfiguration.dockerOptions.containerName, { force: true });
+                    } finally {
+                        disposable.dispose();
+                    }
+                } else {

```

Comment:
Nit: I would just put a comment above the if statement and not use an `else` clause at all. E.g.

`// Don't do anything if this is not our debug session`

Makes things a bit shorter and easier to understand
Your call though